File: robot_catalog_core/src/catalog.cpp
#include "robot_catalog_core/catalog.hpp"
#include "robot_catalog_core/json.hpp"
#include <algorithm>
#include <filesystem>
#include <stdexcept>
#include <yaml-cpp/yaml.h>
#include <ament_index_cpp/get_package_share_directory.hpp>

namespace fs = std::filesystem;

namespace robot_catalog {

static std::string lower(std::string s) {
  std::transform(s.begin(), s.end(), s.begin(), ::tolower);
  return s;
}

static RobotSpecifications parseSpecs(const YAML::Node& n) {
  RobotSpecifications s;
  if (!n) return s;
  s.degrees_of_freedom = n["degrees_of_freedom"].as<int>(0);
  s.payload_kg = n["payload_kg"].as<double>(0.0);
  s.reach_mm = n["reach_mm"].as<double>(0.0);
  s.weight_kg = n["weight_kg"].as<double>(0.0);
  s.repeatability_mm = n["repeatability_mm"].as<double>(0.0);
  s.max_speed_ms = n["max_speed_ms"].as<double>(0.0);
  if (n["mounting"])
    for (const auto& m : n["mounting"]) s.mounting_options.push_back(m.as<std::string>());
  s.safety_certified = n["safety_certified"].as<bool>(false);
  s.collaborative = n["collaborative"].as<bool>(false);
  s.torque_sensing = n["torque_sensing"].as<bool>(false);
  return s;
}

static AttachInfo parseAttach(const YAML::Node& n) {
  AttachInfo a;
  if (!n) return a;
  a.tool_frame = n["tool_frame"].as<std::string>("");
  a.mount_frame = n["mount_frame"].as<std::string>("");
  if (n["xyz"] && n["xyz"].IsSequence())
    for (size_t i = 0; i < 3 && i < n["xyz"].size(); ++i) a.xyz[i] = n["xyz"][i].as<double>(0.0);
  if (n["rpy"] && n["rpy"].IsSequence())
    for (size_t i = 0; i < 3 && i < n["rpy"].size(); ++i) a.rpy[i] = n["rpy"][i].as<double>(0.0);
  return a;
}

static CategoryInfo parseCategory(const std::string& id, const YAML::Node& n) {
  CategoryInfo c;
  c.id = id;
  c.display_name = n["display_name"].as<std::string>("");
  c.description = n["description"].as<std::string>("");
  c.manufacturer = n["manufacturer"].as<std::string>("");
  c.website = n["website"].as<std::string>("");
  return c;
}

static RobotConfig parseRobot(const std::string& id, const YAML::Node& n) {
  RobotConfig r;
  r.id = id;
  r.display_name = n["display_name"].as<std::string>("");
  r.description = n["description"].as<std::string>("");
  r.image_path = n["image_path"].as<std::string>("");
  r.urdf_package = n["urdf_package"].as<std::string>("");
  r.urdf_path = n["urdf_path"].as<std::string>("");
  r.xacro_args = n["xacro_args"].as<std::string>("");
  r.category = n["category"].as<std::string>("");
  r.type = n["type"].as<std::string>("arm");
  r.attach = parseAttach(n["attach"]);
  r.specifications = parseSpecs(n["specifications"]);
  if (n["required_packages"])
    for (const auto& p : n["required_packages"]) r.required_packages.push_back(p.as<std::string>());
  if (n["optional_packages"])
    for (const auto& p : n["optional_packages"]) r.optional_packages.push_back(p.as<std::string>());
  if (n["tags"])
    for (const auto& t : n["tags"]) r.tags.push_back(t.as<std::string>());
  return r;
}

void RobotCatalogCore::mergeFile(const std::string& path) {
  YAML::Node root;
  try {
    root = YAML::LoadFile(path);
  } catch (const std::exception& e) {
    throw std::runtime_error("catalog yaml load failed (" + path + "): " + e.what());
  }
  if (root["categories"]) {
    for (const auto& kv : root["categories"]) {
      std::string id = kv.first.as<std::string>();
      categories_[id] = parseCategory(id, kv.second);
    }
  }
  if (root["robots"]) {
    for (const auto& kv : root["robots"]) {
      std::string id = kv.first.as<std::string>();
      robots_[id] = parseRobot(id, kv.second);
    }
  }
}

void RobotCatalogCore::load(const std::string& path) {
  std::error_code ec;
  if (fs::is_directory(path, ec)) {
    loadFromDirectory(path);
  } else {
    loadFromFile(path);
  }
}

void RobotCatalogCore::loadFromFile(const std::string& path) {
  robots_.clear();
  categories_.clear();
  mergeFile(path);
  rebuildCaches();
}

void RobotCatalogCore::loadFromDirectory(const std::string& dir) {
  robots_.clear();
  categories_.clear();
  std::vector<fs::path> files;
  for (const auto& entry : fs::recursive_directory_iterator(dir)) {
    if (!entry.is_regular_file()) continue;
    auto ext = entry.path().extension().string();
    if (ext == ".yml" || ext == ".yaml") files.push_back(entry.path());
  }
  // Deterministic merge order so later files override earlier ones predictably.
  std::sort(files.begin(), files.end());
  for (const auto& f : files) mergeFile(f.string());
  rebuildCaches();
}

void RobotCatalogCore::rebuildCaches() {
  sorted_robots_.clear();
  sorted_robots_.reserve(robots_.size());
  for (const auto& kv : robots_) sorted_robots_.push_back(kv.second);
  std::sort(sorted_robots_.begin(), sorted_robots_.end(),
            [](const RobotConfig& a, const RobotConfig& b) { return a.id < b.id; });
  all_robots_json_ = robots_to_json(sorted_robots_);
  categories_json_ = categories_to_json(getCategories());
}

std::vector<RobotConfig> RobotCatalogCore::getAllRobots() const {
  return sorted_robots_;
}

std::vector<RobotConfig> RobotCatalogCore::getRobotsPage(int offset, int limit,
                                                         int& total) const {
  total = static_cast<int>(sorted_robots_.size());
  std::vector<RobotConfig> out;
  if (offset < 0) offset = 0;
  if (offset >= total) return out;
  int end = (limit <= 0) ? total : std::min(total, offset + limit);
  for (int i = offset; i < end; ++i) out.push_back(sorted_robots_[i]);
  return out;
}

bool RobotCatalogCore::getRobotById(const std::string& id, RobotConfig& out) const {
  auto it = robots_.find(id);
  if (it == robots_.end()) return false;
  out = it->second;
  return true;
}

std::vector<CategoryInfo> RobotCatalogCore::getCategories() const {
  std::vector<CategoryInfo> out;
  for (const auto& kv : categories_) out.push_back(kv.second);
  std::sort(out.begin(), out.end(),
            [](const CategoryInfo& a, const CategoryInfo& b) { return a.id < b.id; });
  return out;
}

std::vector<RobotConfig> RobotCatalogCore::getRobotsByCategory(const std::string& category) const {
  std::vector<RobotConfig> out;
  for (const auto& r : sorted_robots_)
    if (r.category == category) out.push_back(r);
  return out;
}

bool RobotCatalogCore::matchesFilter(const RobotConfig& r, const RobotFilter& f) {
  const auto& s = r.specifications;
  if (f.type && r.type != *f.type) return false;
  if (f.category && r.category != *f.category) return false;
  if (f.min_payload && s.payload_kg < *f.min_payload) return false;
  if (f.max_payload && s.payload_kg > *f.max_payload) return false;
  if (f.min_reach && s.reach_mm < *f.min_reach) return false;
  if (f.max_reach && s.reach_mm > *f.max_reach) return false;
  if (f.degrees_of_freedom && s.degrees_of_freedom != *f.degrees_of_freedom) return false;
  if (f.collaborative_only && *f.collaborative_only && !s.collaborative) return false;
  for (const auto& tag : f.required_tags)
    if (std::find(r.tags.begin(), r.tags.end(), tag) == r.tags.end()) return false;
  if (!f.search_text.empty() && !matchesSearch(r, f.search_text)) return false;
  return true;
}

bool RobotCatalogCore::matchesSearch(const RobotConfig& r, const std::string& term) {
  std::string needle = lower(term);
  std::vector<std::string> hay = {r.display_name, r.description, r.category};
  hay.insert(hay.end(), r.tags.begin(), r.tags.end());
  for (auto& h : hay)
    if (lower(h).find(needle) != std::string::npos) return true;
  return false;
}

std::vector<RobotConfig> RobotCatalogCore::filterRobots(const RobotFilter& f) const {
  if (f.isEmpty()) return getAllRobots();
  std::vector<RobotConfig> out;
  for (const auto& r : sorted_robots_)
    if (matchesFilter(r, f)) out.push_back(r);
  return out;
}

std::vector<RobotConfig> RobotCatalogCore::searchRobots(const std::string& term) const {
  if (term.empty()) return getAllRobots();
  std::vector<RobotConfig> out;
  for (const auto& r : sorted_robots_)
    if (matchesSearch(r, term)) out.push_back(r);
  return out;
}

std::vector<std::string> RobotCatalogCore::getMissingPackages(const RobotConfig& r) const {
  std::vector<std::string> missing;
  for (const auto& pkg : r.required_packages) {
    try {
      ament_index_cpp::get_package_share_directory(pkg);
    } catch (const std::exception&) {
      missing.push_back(pkg);
    }
  }
  return missing;
}

bool RobotCatalogCore::validateRobotPackages(const RobotConfig& r) const {
  return getMissingPackages(r).empty();
}

}  // namespace robot_catalog

File: web/backend/rdsa_web/catalog.py
"""Robot catalog: loads robots.yaml into models and provides queries."""
from __future__ import annotations

from pathlib import Path

import yaml

from .models import (
    AttachInfo,
    CategoryInfo,
    RobotConfig,
    RobotFilter,
    RobotSpecifications,
)


class RobotCatalog:
    def __init__(
        self,
        robots: dict[str, RobotConfig],
        categories: dict[str, CategoryInfo],
    ) -> None:
        self._robots = robots
        self._categories = categories

    # -- construction --------------------------------------------------
    @classmethod
    def from_file(cls, path: str | Path) -> "RobotCatalog":
        """Load a single YAML file, or every *.yml/*.yaml under a directory."""
        p = Path(path)
        if p.is_dir():
            return cls.from_directory(p)
        return cls._from_data([yaml.safe_load(p.read_text()) or {}])

    @classmethod
    def from_directory(cls, path: str | Path) -> "RobotCatalog":
        p = Path(path)
        files = sorted(set(p.rglob("*.yml")) | set(p.rglob("*.yaml")))
        return cls._from_data([yaml.safe_load(f.read_text()) or {} for f in files])

    @classmethod
    def _from_data(cls, datas: list[dict]) -> "RobotCatalog":
        categories: dict[str, CategoryInfo] = {}
        robots: dict[str, RobotConfig] = {}
        for data in datas:
            for cid, node in (data.get("categories") or {}).items():
                categories[cid] = cls._parse_category(cid, node)
            for rid, node in (data.get("robots") or {}).items():
                robots[rid] = cls._parse_robot(rid, node)
        return cls(robots, categories)

    @staticmethod
    def _parse_category(cid: str, node: dict) -> CategoryInfo:
        return CategoryInfo(
            id=cid,
            display_name=node.get("display_name", ""),
            description=node.get("description", ""),
            manufacturer=node.get("manufacturer", ""),
            website=node.get("website", ""),
        )

    @staticmethod
    def _parse_specs(node: dict | None) -> RobotSpecifications:
        node = node or {}
        return RobotSpecifications(
            degrees_of_freedom=node.get("degrees_of_freedom", 0),
            payload_kg=node.get("payload_kg", 0.0),
            reach_mm=node.get("reach_mm", 0.0),
            weight_kg=node.get("weight_kg", 0.0),
            repeatability_mm=node.get("repeatability_mm", 0.0),
            max_speed_ms=node.get("max_speed_ms", 0.0),
            mounting_options=list(node.get("mounting", []) or []),
            safety_certified=node.get("safety_certified", False),
            collaborative=node.get("collaborative", False),
            torque_sensing=node.get("torque_sensing", False),
        )

    @staticmethod
    def _parse_attach(node: dict | None) -> AttachInfo:
        node = node or {}
        return AttachInfo(
            tool_frame=node.get("tool_frame", ""),
            mount_frame=node.get("mount_frame", ""),
            xyz=list(node.get("xyz", [0.0, 0.0, 0.0]) or [0.0, 0.0, 0.0]),
            rpy=list(node.get("rpy", [0.0, 0.0, 0.0]) or [0.0, 0.0, 0.0]),
        )

    @classmethod
    def _parse_robot(cls, rid: str, node: dict) -> RobotConfig:
        return RobotConfig(
            id=rid,
            display_name=node.get("display_name", ""),
            description=node.get("description", ""),
            image_path=node.get("image_path", ""),
            urdf_package=node.get("urdf_package", ""),
            urdf_path=node.get("urdf_path", ""),
            xacro_args=node.get("xacro_args", ""),
            category=node.get("category", ""),
            type=node.get("type", "arm"),
            attach=cls._parse_attach(node.get("attach")),
            specifications=cls._parse_specs(node.get("specifications")),
            required_packages=list(node.get("required_packages", []) or []),
            optional_packages=list(node.get("optional_packages", []) or []),
            tags=list(node.get("tags", []) or []),
        )

    # -- queries -------------------------------------------------------
    def get_all_robots(self) -> list[RobotConfig]:
        return sorted(self._robots.values(), key=lambda r: r.id)

    def get_robots_page(
        self, offset: int = 0, limit: int = 0
    ) -> "tuple[list[RobotConfig], int]":
        """Stable page of all robots; limit <= 0 means all. Returns (page, total)."""
        robots = self.get_all_robots()
        total = len(robots)
        offset = max(offset, 0)
        end = total if limit <= 0 else min(total, offset + limit)
        return robots[offset:end], total

    def get_robot_by_id(self, robot_id: str) -> RobotConfig | None:
        return self._robots.get(robot_id)

    def get_categories(self) -> list[CategoryInfo]:
        return list(self._categories.values())

    def get_category_info(self, category_id: str) -> CategoryInfo | None:
        return self._categories.get(category_id)

    def get_robots_by_category(self, category: str) -> list[RobotConfig]:
        return [r for r in self._robots.values() if r.category == category]

    def filter_robots(self, flt: RobotFilter) -> list[RobotConfig]:
        if flt.is_empty():
            return self.get_all_robots()
        return [r for r in self.get_all_robots() if self._matches_filter(r, flt)]

    def filter_robots_page(
        self, flt: RobotFilter, offset: int = 0, limit: int = 0
    ) -> "tuple[list[RobotConfig], int]":
        matched = self.filter_robots(flt)
        total = len(matched)
        offset = max(offset, 0)
        end = total if limit <= 0 else min(total, offset + limit)
        return matched[offset:end], total

    @staticmethod
    def _matches_filter(r: RobotConfig, flt: RobotFilter) -> bool:
        s = r.specifications
        if flt.type is not None and r.type != flt.type:
            return False
        if flt.category is not None and r.category != flt.category:
            return False
        if flt.min_payload is not None and s.payload_kg < flt.min_payload:
            return False
        if flt.max_payload is not None and s.payload_kg > flt.max_payload:
            return False
        if flt.min_reach is not None and s.reach_mm < flt.min_reach:
            return False
        if flt.max_reach is not None and s.reach_mm > flt.max_reach:
            return False
        if (
            flt.degrees_of_freedom is not None
            and s.degrees_of_freedom != flt.degrees_of_freedom
        ):
            return False
        if flt.collaborative_only is True and not s.collaborative:
            return False
        if any(tag not in r.tags for tag in flt.required_tags):
            return False
        if flt.search_text and not RobotCatalog._matches_search(r, flt.search_text):
            return False
        return True

    def search_robots(self, term: str) -> list[RobotConfig]:
        if not term:
            return self.get_all_robots()
        return [r for r in self._robots.values() if self._matches_search(r, term)]

    @staticmethod
    def _matches_search(r: RobotConfig, term: str) -> bool:
        needle = term.lower()
        haystacks = [r.display_name, r.description, r.category, *r.tags]
        return any(needle in h.lower() for h in haystacks)

    # -- URDF / mesh / validation ----------------------------------------
    # Pure-Python fallback used only in no-ROS/test mode. In production the
    # C++ catalog backend (CppCatalog) performs all of this work; this class
    # mirrors that interface so app.py can stay backend-agnostic.
    def get_urdf(self, robot_id: str) -> dict:
        from .urdf import UrdfError, resolve_urdf
        from .validation import get_missing_packages

        robot = self.get_robot_by_id(robot_id)
        if robot is None:
            return {"found": False, "ok": False, "urdf_xml": "",
                    "missing_packages": [], "error": ""}
        missing = get_missing_packages(robot)
        try:
            xml = resolve_urdf(robot)
            return {"found": True, "ok": True, "urdf_xml": xml,
                    "missing_packages": missing, "error": ""}
        except UrdfError as exc:
            return {"found": True, "ok": False, "urdf_xml": "",
                    "missing_packages": missing, "error": str(exc)}

    def resolve_mesh(self, package: str, rel_path: str) -> "tuple[bytes, str] | None":
        import mimetypes

        from .catalog_client import MESH_SIZE_CAP, MeshTooLarge
        from .meshes import resolve_mesh_path

        path = resolve_mesh_path(package, rel_path)
        if path is None:
            return None
        size = path.stat().st_size
        if size > MESH_SIZE_CAP:
            raise MeshTooLarge(size)
        media_type = mimetypes.guess_type(str(path))[0] or "application/octet-stream"
        return path.read_bytes(), media_type

    def validate(self, robot_id: str) -> dict:
        from .validation import get_missing_packages

        robot = self.get_robot_by_id(robot_id)
        if robot is None:
            return {"found": False, "ok": False, "missing_packages": []}
        missing = get_missing_packages(robot)
        return {"found": True, "ok": not missing, "missing_packages": missing}
